### ISSUE TITLE
Meson compatibility policy

### ISSUE THREAD

**Issue by dnicolodi** [mesonbuild/meson-python#392]
The minimum Meson version currently required by meson-python is 0.63.3 because of some fixes included in that release that make some fundamental aspect of meson-python to work. Since we bumped the requirement a few months ago, there have been a few releases of Meson but nothing impacting meson-python in particular. Thus the minimum required version has not been bumped. We have CI jobs that try to verify that we don't introduce changes to meson-python that break support for older Meson versions. We test with Meson 0.63.3, 0.64.1, and 1.0.1 in addition to the latest 1.1.0 release, in this moment.

A few weeks ago a couple of fixes for Meson bugs affecting meson-python have been merged: https://github.com/mesonbuild/meson/pull/11655 https://github.com/mesonbuild/meson/pull/11599 These have not been part of any release, but they have been tagged for inclusion in a possible Meson 1.0.2 release.

For the first bug, we currently carry a work around. The second bug affects projects that use the Meson `install_data()` function with the `exclude_directories` and `exclude_files`: these do not work correctly in Windows, see #352.

What should we do when these fixes will be included in a Meson release? Can we bump the required Meson version to 1.0.2 o 1.1.1?

Is there are any drawback in "aggressively" bumping the required Meson version? I think the only use case negatively affected are distributions, where the latest Meson mersion may take a while to be packaged. Do we know of Linux distributions packaging meson-python or packages having meson-python as a build dependency? It seems that Debian has a package for meson-python. What is the situation for the conda ecosystem?

**Comment by rgommers** [comment]
I think it's fine to be fairly aggressive with updates. Linux distros that package `meson-python` and `meson` can upgrade them in sync in principle. The only potential issue with that is if a new `meson` version would contain a regression that prevents another package from building. This is true for regular Python usage from PyPI as well, and I think it's the main downside: regressions are harder to work around. Usage of `git bisect` may also be annoying if the version range of build dependencies is very narrow.

My feeling is:
- supporting the latest ~3 Meson minor releases (= a 6-9 month window) should be enough
- if we have a good reason, we can bump the minimum version more aggressively
- in this case, we're probably okay with leaving it as is, given that (a) the bugs fixed will only affect a limited number of users, and (b) for regular `requires = ['meson-python']` usage, the highest Meson version is pulled in anyway.

> What is the situation for the conda ecosystem?

Conda-forge packages both `meson` and `meson-python` quite quickly. There's nothing to worry about there.

**Comment by eli-schwartz** [comment]
> I think the only use case negatively affected are distributions, where the latest Meson mersion may take a while to be packaged. Do we know of Linux distributions packaging meson-python or packages having meson-python as a build dependency? It seems that Debian has a package for meson-python.

Linux distros will tend to refuse updating either meson or meson-python, because stability policies, or else update both at the same time.

However, users might install meson from the distro and then later pip install something that needs meson-python, so it's still possible to end up with slightly older meson, and you might want to support that...

This probably only matters for people disabling build isolation anyway? (Because with build isolation the latest version of meson is always installed from scratch.) Not very common on Windows, in particular, I'd imagine.

**Comment by musicinmybrain** [comment]
I just introduced a [`python-meson-python` package](https://src.fedoraproject.org/rpms/python-meson-python) to Fedora Linux.

We [release a new version of Fedora Linux approximately every six months and provide updated packages (maintenance) to these releases for approximately 13 months](https://docs.fedoraproject.org/en-US/releases/lifecycle/). Our [Updates Policy](https://docs.fedoraproject.org/en-US/fesco/Updates_Policy/) would allow us to update meson in a stable release as long as the new version didn’t have breaking changes—so either bugfix/patch releases or compatible feature releases are OK—but in practice, it’s up to the package’s maintainers how aggressively they want to validate and issue compatible updates in stable releases, and validation is a greater effort for packages that are as widely used as meson.

You can see the current status of meson in Fedora [here](https://src.fedoraproject.org/rpms/meson). Note that F38 was just released, and F36 is about to reach end-of-life. I would expect a particular Fedora Linux version is likely to stay on a particular meson feature release (e.g. 1.0.x vs. 1.1.x) unless there is an unusually good reason to update. There are a lot of packages that depend on meson, and a lot of ways they can break, including ways that don’t cause them to fail to build from source. Imagine the pkgconfig module in meson had a regression and `.pc` files were invalid; you might not see it until you tried to rebuild a package that *depended* on a package that was rebuilt with the new meson. Still, since meson is an important package with active maintainers, you can expect that at least Rawhide (the development version of Fedora Linux) will have the latest meson fairly soon after it is released.

When it comes to meson-python, as the primary package maintainer for the Fedora package, I will take updates in stable releases on a case-by-case basis for now. I’m inclined to keep these updated pretty aggressively at this early stage of meson-python development, as long as the meson version is new enough, the meson-python release notes don’t call out any obvious breaking changes, and the dependent packages still build from source.

When an eventual [EPEL10](https://docs.fedoraproject.org/en-US/epel/) comes around, if `python-meson-python` is branched for it, you can expect that meson will stay at essentially the same version forever in RHEL10, and so `python-meson-python` will eventually be blocked by the meson version or the version of some other dependency and become frozen in time in EPEL10 as well. Such is the nature of enterprise distributions.

I hope that helps. Feel free to mention me if there is something specific you would like to discuss related to Fedora packaging.

**Comment by eli-schwartz** [comment]
@musicinmybrain, do you happen to know if RHEL did anything with the meson 0.61.x release series?

Typically meson has new feature releases once every few months, and then bugfix releases on a maintenance branch in between then and the next feature release. A notable exception was 0.61, the last feature release to include support for python 3.6 before its End of Life date. A redhat employee working on software that builds with meson actually did contribute quite a bit of time to prepare patchsets that backported bugfixes to that release branch, so meson ended up releasing LTS bugfixes all the way to a .5 release... however I'm not sure that he had anything to do with meson packaging. ;)

(This offer is generally open. If people want it and we've stopped actively maintaining a release, they are welcome to nominate backport-worthy patches and resolve any possible cherry-pick conflicts, and we can coordinate official releases.)

**Comment by musicinmybrain** [comment]
> @musicinmybrain, do you happen to know if Fedora/RHEL did anything with the meson 0.61.x release series?

I can’t conjecture what that Red Hat employee was working on (I’m a Fedora volunteer, not a Red Hat employee), but I can report package versions. Fedora 33 had 0.59.4 when it went EOL on 2021-11-30, and Fedora 34 had 0.62.1 by the time it went EOL on 2022-06-07. RHEL9 and RHEL8 both currently have 0.58.2. Python 3.6 is the system Python on RHEL8. Based on all of that, it doesn’t *look* like the 0.61.x series got any special treatment from the packaging side.

**Comment by eli-schwartz** [comment]
I see. Thanks for the info. I guess it would be kind of nice to have 0.61.5 backported but idk how that works or who to ask, so be it...
